tensorcircuit/mpscircuit.py

```diff
@@ -532,83 +763,180 @@ def proj_with_mps(self, other: "MPSCircuit") -> Tensor:
+        ):
+            center_position = self._mps.center_position - begin
+
+        mps = self.__class__(
+            nqubits,
+            tensors=tensors,
+            center_position=center_position,
+            **self.rules,
+        )
+        return mps
+
+    def expectation(
+        self,
+        *ops: Tuple[Gate, List[int]],
+        resue: bool = True,
+        other: Optional["MPSCircuit"] = None,
+        conj: bool = True,
+        normalize: bool = False,
+        **rules: Any,
+    ) -> Tensor:
         """
         Compute the expectation of corresponding operators in the form of tensor.
 
         :param ops: Operator and its position on the circuit,
             eg. ``(gates.Z(), [1]), (gates.X(), [2])`` is for operator :math:`Z_1X_2`
         :type ops: Tuple[tn.Node, List[int]]
+        :param reuse: If True, then the wavefunction tensor is cached for further expectation evaluation,
+            defaults to be true.
+        :type reuse: bool, optional
+        :param other: If not None, will be used as bra
+        :type other: MPSCircuit, optional
+        :param conj: Whether to conjugate the bra state
+        :type conj: bool, defaults to be True
+        :param normalize: Whether to normalize the MPS
+        :type normalize: bool, defaults to be True
+        :param rules: Truncation rules
+        :type rules: Any
         :return: The expectation of corresponding operators
         :rtype: Tensor
         """
-        # A better idea is to create a MPO class and have a function to transform gates to MPO
-        mpscircuit = self.copy()
+        # If the bra is ket itself, the environments outside the operators can be viewed as identities,
+        # so does not need to contract
+        ops = [list(op) for op in ops]  # type: ignore # turn to list for modification
+        for op in ops:
+            if isinstance(op[1], int):
+                op[1] = [op[1]]
+        all_sites = np.concatenate([op[1] for op in ops])
+
+        # move position inside the operator range
+        if other is None:
+            site_begin = np.min(all_sites)
+            site_end = np.max(all_sites)
+            if self._mps.center_position < site_begin:
+                self.position(site_begin)
+            elif self._mps.center_position > site_end:
+                self.position(site_end)
+        else:
+            assert isinstance(other, MPSCircuit), "the bra has to be a MPSCircuit"
+
+        # apply the gate
+        mps = self.copy()
+        # when the truncation rules is None (which is the default),
+        # it is guaranteed that the result is a real number
+        # since the calculation is exact, otherwise there's no guarantee
+        mps.set_truncation_rule(**rules)
         for gate, index in ops:
-            mpscircuit.apply_general_gate(gate, *index)
-        value = mpscircuit.proj_with_mps(self)
-        return backend.convert_to_tensor(value)
+            mps.apply(gate, *index)
```

Comment:
these gate in ops can be non-unitary, pls double check this is ok here

Reply:
Unfortunately all gates are assumed to be unitary throughout the developing... So probably need some significant changes later. Could be added into the future steps?

Reply:
The issue is that the unitary assumption sometimes offer additional benefit so maybe we need some label (probably need to specified by the user?) of whether a gate is unitary or not?

Reply:
Here the problem is not supporting nonunitary gate in the circuit (not on the top of  priority list), here the gate is actually op for expectation which can be non-unitary very easily

Reply:
Good point. Didn't consider this at all but seems super important. After some thinking, only single-qubit gate is affected because applying unitary single gate does not need to move the canonicalization center, all others are not affected, so should be easy to fix.